engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java

```diff
@@ -41,13 +45,28 @@
 	private final ConstraintLocation location;
 
 	/**
+	 * The {@link ValueExtractor} used to extract the value for validation.
+	 */
+	private final ValueExtractorDescriptor valueExtractorDescriptor;
+
+	/**
+	 * The type of the validated element.
+	 */
+	private final Type typeOfValidatedElement;
```

Comment:
Yeah... The fact is that it's currently very hard to understand what exactly the various types are for. I kept the behavior as it was before. It may be suboptimal but I supposed the existing code was here for a reason. If you want to disambiguate them, I would appreciate some explanation/comments because I was not able to do it myself.